### PR TITLE
BUG #10228: segfault due to out-of-bounds in binning

### LINKED ISSUE #10228: resample() with how=count causes Segmentation Fault
Runing the code as is below results in a segmentation fault when it gets to resample()
The first pass runs, and then either the 2nd or 3rd run produces:

Segmentation fault: 11

It happens on:
- Python 2.7.9 and Pandas 0.16.1
- Python 3.4.3 and Pandas 0.16.1

It does not happen on:
- Python 2.7.9 with Pandas 0.12.0

Those are the only ones I've tested

```
import pandas as pd

frame = pd.DataFrame({'delay_category': {pd.Timestamp('2014-05-18 08:50:00'): 'yes',
  pd.Timestamp('2014-05-19 14:35:00'): 'yes',
  pd.Timestamp('2014-05-22 08:45:00'): 'yes',
  pd.Timestamp('2014-05-27 10:00:00'): 'no',
  pd.Timestamp('2014-05-28 11:05:00'): 'no'}})

frame_filter = frame[frame['delay_category'] == 'da']['delay_category']

for i in range(1,10):
    frame_filter.resample('d', how='count')

```
Setting how=sum also does not result in the issue.

It only happens after the frame has been selected with a parameter that doesn't work, resulting in an empty series.
Eg. this works fine:
```
frame_filter = frame[frame['delay_category'] == 'yes']['delay_category']

for i in range(1,10):
    frame_filter.resample('d', how='count')

```

### PR DESCRIPTION
**Pull request by Garrett-R**
Closes #10228.   I also deleted some duplicated code while I was at it.

So, I wasn't sure if I should be including a unit test for this.  This issue was a segfault, which was happening when you do:

```
s = pd.Series([], index=pd.DatetimeIndex([]), dtype=np.object)
s.resample('d', how='count')
```

so I suppose I could add this code into a test, and just verify it doesn't segfault, but that seemed like a bad idea.  It would be testing for something that's undefined behavior and therefore be a non-deterministic test.

### CODE REVIEW

**Comment by jreback** [comment]
of course you need a test, otherwise how can you tell if your fix works? The test can be as simple as asserting that it runs.

File: pandas/src/generated.pyx [hunk 1]
```diff
@@ -9110,6 +9110,8 @@ def group_count_bin_float64(ndarray[float64_t, ndim=2] out,
         ndarray[int64_t, ndim=2] nobs = np.zeros((out.shape[0], out.shape[1]),
                                                  dtype=np.int64)
 
+    if len(bins) == 0:
```
Comment (jreback) [inline pandas/src/generated.pyx]: this is by definition **generated** code, edits here will be lost. rather make the fix in ``src/generate_code.py``, then``python generate_code.py`` creates the ``generated.pyx``

Comment (Garrett-R) [inline pandas/src/generated.pyx]: Oh crap, my bad!  I did see the name, but then I looked at git history and saw what appeared to be people editing it (but I get it now).

Do you think it's worth adding a warning message at the top for new devs [like NumPy does](https://github.com/numpy/numpy/blob/master/numpy/distutils/conv_template.py#L57)?  Or should it be obvious from the name?

**Comment by Garrett-R** [comment]
@jreback, sure I'll write a test.  I have a piece of code that reliably segfaults for me:

    ss = pd.Series([], index=pd.DatetimeIndex([]), dtype=np.object)
    ss.resample('d', how='count')

That being said, this doesn't mean it'll reliably segfault for other machines since out-of-bounds access are undefined behavior.  One way would be to re-enable the bounds-checking on these Cython functions just for the unit tests, but maybe that's over-thinking it...

**Comment by jreback** [comment]
@Garrett-R then that is a reasonable test. Even if it doesn't segfault on some platforms that is ok. Note that you *can* change the directive at run-time, see [here](http://docs.cython.org/src/reference/compilation.html#how-to-set-directives), e.g. you would have a flag that is passed in to re-enable this. But to be honest it is overkill, and not necessary to tests with that. Since you have an example that fails, that can serve as a smoke tests (e.g. the test is simply running that code, if it doesn't segfault then the test passes).

**Comment by jreback** [comment]
@Garrett-R No objection to a separate PR for adding a warning to ``generated.pyx`` (and more importantly prob a note in ``internals.rst`` about how/what to change. thanks!

File: pandas/tseries/tests/test_resample.py [hunk 2]
```diff
@@ -660,6 +660,20 @@ def test_resample_empty(self):
         rs = xp.resample('A')
         assert_frame_equal(xp, rs)
 
+        # These were sometimes causing a segfault (for the functions with
+        # bounds-checking disabled) or an IndexError.  We just run them to
+        # ensure they no longer do.  (#10228)
+        index = pd.DatetimeIndex([])
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: I would actually make a couple of loops here, iterate over all of the index types, then all of the possible dtypes, then all of the 'hows' and exhaustivly test the possibilities

I would actually define these functions in ``utils.testing`` (right below the make index) functions
```
def all_index_types():
  return [ tm.makeIntIndex, tm.makeFloatIndex,
               tm.makeStringIndex, tm.makeUnicodeIndex,
                tm.makeDateIndex, tm.makePeriodIndex, 
                tm.makeTimedeltaIndex, tm.makeBoolIndex, 
                tm.makeCategoricalIndex]

def datetimelike_index_types():
     return [  tm.makeDateIndex, tm.makePeriodIndex, 
                tm.makeTimedeltaIndex ]
```

obviously use the 2nd one

**Comment by Garrett-R** [comment]
@jreback, good call about adding more comprehensive testing.  This helped me find some more out-of-bounds bugs (which were in `algos.pyx`).

I'll send a PR soon to supply a "DO NOT EDIT THIS GENERATED FILE...." warning for other fellow noobs.

File: pandas/tseries/tests/test_resample.py (hunk 2 above)
Comment (Garrett-R) [inline pandas/tseries/tests/test_resample.py]: Cool, I ended up making these generators that returned the index instances, which I thought would keep the test cleaner.  Let me know if you prefer to just return a list of functions though and I can change it.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/25fc49d975d74c5cb969089dcb3efeda15c3262f

thanks!

nice fix.

btw if you'd like to update the docs w.r.t. the ``generate_code.pyx`` would be gr8. pls put in ``internals.rst``

**Pull request closed without merging**